tutorials/git-advanced.rst

```diff
@@ -0,0 +1,604 @@
+    =======
+        char input[1000];
+        while(fgets(input, 1000, stdin)){  //read from STDIN (aka command-line)
+    >>>>>>> update-buffer-size
+
+with this:
+
+::
+
+    char input[300];
+    while(fgets(input, 300, stdin)){  //read from STDIN (aka command-line)
+
+i.e., the full ``echo.c`` file should look like this::
+
+    /* echo.c - echo user input */
+    /* YOUR_NAME */
+    /* Last updated: 3/28/22 */
+
+    #include <stdio.h>
+    #include <stdlib.h>
+    #include <string.h>
+
+    int main(){
+        char input[300];
+        while(fgets(input, 300, stdin)){  //read from STDIN (aka command-line)
+            printf("%s\n", input);  //print out what user typed in
+            memset(input, 0, strlen(input));  //reset string to all 0's
+        }
+        return 1;
+    }
+
+Now, we need to tell Git that we’ve resolved the merge conflict. We do
+this by using ``git add``:
+
+::
+
+   git add echo.c
+
+Note how now ``git status`` recognizes that the merge conflict has been
+resolved, but has yet to be committed:
+
+::
+
+    $ git status
+    On branch main
+    Your branch is ahead of 'origin/main' by 2 commits.
+      (use "git push" to publish your local commits)
+
+    All conflicts fixed but you are still merging.
+      (use "git commit" to conclude merge)
+
+    Changes to be committed:
+        modified:   echo.c
+
+
+So, all we need to do is to run this command:
+
+::
+
+   git commit
+
+This will open up an editor with a default commit message like
+``Merge branch 'update-buffer-size' into main``. You could change this
+to something like ``Merging 'update-buffer-size' (buffer should actually be 300)``
+to indicate that you did not actually accept the changes from the ``update-buffer-size``
+branch but, for the purposes of this tutorial, you can also just use the default message.
+
+Once you save the commit message, the
+merge will be completed and you will see something like this:
+
+::
+
+   [main e58a1ba] Merge branch 'update-buffer-size' into main
+
+That is the *merge commit* for this merge; if you run ``git log``, you
+will see that the commit history now includes the commit from
+``update-buffer-size``, as well as the merge commit. Notice how it includes a
+``Merge:`` line telling us which two commits were merged:
+
+::
+
+    commit e58a1baa2d6d408027a04d19ca322ef4ceaae9da (HEAD -> main)
+    Merge: e5ec414 a72bbd0
+    Author: Borja Sotomayor <borja@cs.uchicago.edu>
+    Date:   Sat Mar 27 11:25:43 2021 -0500
+
+        Merge branch 'update-buffer-size' into main
+
+    commit e5ec414fb5422487f3d0469583461c9a260432d9
+    Author: Borja Sotomayor <borja@cs.uchicago.edu>
+    Date:   Sat Mar 27 11:15:04 2021 -0500
+
+        Updated buffer size to 250
+
+    commit a72bbd0ee5302906177cc9f62d4ff55e1a02c999 (update-buffer-size)
+    Author: Borja Sotomayor <borja@cs.uchicago.edu>
+    Date:   Sat Mar 27 11:12:57 2021 -0500
+
+        Updated buffer size to 1000
+
+
```

Comment:
Having taken CS 220 I learned about CI and it made my life a lot easier. I would suggest adding a section not necessarily on CI per se, but on the ideas behind it. Telling students about branches doesn't mean they'll know how to use them effectively. I think it would be good to tell them to use branches in conjunction with frequent and isolated changes.

Reply:
Please create an issue related to adding a section on CI. Like the section on pull requests, I think this is a good idea, but will take a bit more time to do.